File: 백준/Silver/1189. 컴백홈/컴백홈.cc
#include <iostream>
#include <algorithm>
using namespace std;

int n, m, l, res;
char arr[5][5];
int visit[5][5];

int dr[4] = { 1, 0, -1, 0 };
int dc[4] = { 0, 1, 0, -1 };

void dfs(int r, int c, int cnt) {
	if (r == 0 && c == m - 1) {
		if (cnt == l) res++;
		return;
	}

	for (int i = 0; i < 4; i++) {
		int rr = r + dr[i];
		int cc = c + dc[i];

		if (rr < 0 || cc < 0 || rr >= n || cc >= m) continue;
		if (arr[rr][cc] == 'T' || visit[rr][cc]) continue;

		visit[rr][cc] = true;
		dfs(rr, cc, cnt + 1);
		visit[rr][cc] = false;
	}
}

int main() {
	cin >> n >> m >> l;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++) cin >> arr[i][j];

	visit[n - 1][0] = true;
	dfs(n - 1, 0, 1);

	cout << res;
}

File: 백준/Silver/1189. 컴백홈/컴백홈.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int n;
    static int m;
    static int l;
    static int res = 0;

    static int[] DR = {0, 1, 0, -1};
    static int[] DC = {1, 0, -1, 0};

    static char[][] arr;
    static boolean[][] vis;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        l = Integer.parseInt(st.nextToken());

        arr = new char[n][m];
        vis = new boolean[n][m];

        for (int i = 0; i < n; i++){
            arr[i] = br.readLine().toCharArray();
        }

        vis[n - 1][0] = true;

        dfs(n - 1, 0, 1);

        System.out.println(res);
    }

    public static void dfs(int r, int c, int cnt){

        if (r == 0 && c == m - 1) {
            if (cnt == l)
                res++;
            return;
        }

        for (int i = 0; i < 4; i++) {
            int nextR = r + DR[i];
            int nextC = c + DC[i];

            if (nextR < 0 || nextC < 0 || nextR >= n || nextC >= m)
                continue;

            if (arr[nextR][nextC] == 'T' || vis[nextR][nextC])
                continue;

            vis[nextR][nextC] = true;

            dfs(nextR, nextC, cnt + 1);

            vis[nextR][nextC] = false;
        }
    }
}